README.md
<br />

<p align="middle">
    <img src="https://github.com/antz22/retro/blob/master/screenshots/landing.png" width="100%" style="margin:0; padding: 0">
</p>

<br />

<p align="middle">
  Retro is an industry-changing application that provides valuable information to first responders in the case of emergency and decentralizes the storage of medical records. 
</p>


# Inspiration
Our inspiration for this project was twofold. First, we found that, during our education in health and first-aid, many tragedies and accidents could have been easily prevented if the first responders had greater access to the medical history of the patient. For instance, knowledge of a simple allergy or a serious leg injury would greatly guide the care of a physician. Second, in the realm of healthcare, many hospitals take advantage of patients by obstructing patients from gaining access to their medical records by charging expensive fees. By decentralizing the storage of medical records and giving more power to the patient, Retro is able to improve access to healthcare for minorities who may find it more difficult to pay fees and observe crucial information on their medical history.

# What it does
Retro is a multi-faceted app that extends to both users and observers. Users register to create a unique profile of their medical history and information, such as age, sex, height, weight, and more. Then, throughout their lifetime, users can add information to the app, including information on conditions, treatments, and blood tests. In a beautiful dashboard on the web app, users can see insights on their medical history and information. More importantly, however, the app generates unique qr codes for every user that, in the case of an emergency, can be scanned by first responders to gain more insights on how best to treat the patient. Both users and doctors can also upload important PDF files to keep track of past X-Ray scans, patient files, and more, in one place.

<p align="middle">
    <img src="https://github.com/antz22/retro/blob/master/screenshots/dashboard.png" width="100%" style="margin:0; padding: 0">
</p>

# How we built it
To build our app, we used HTML and CSS on frontend in conjunction with a Python Flask backend that interacts with a Firebase Realtime Database. We also used the MapBox API and QRCode library to generate the QR Codes and provide privacy checks for only people in the perimeter of the patient to scan their qrcode.

# Challenges we ran into
During our project, we faced several challenges with finding a suitable Map API that could determine the proximity of nearby hospitals based on latitude and longitude. We also had a difficult time in using the Python Flask library in conjunction with Firebase, a combination of technologies that was new to all of us.

# Accomplishments that we're proud of
We are particularly proud of the UI design and high level of customizability of our medical portal and dashboard. The dashboard provides numerous insights onto the user’s medical history. Additionally, the added level of security with the Map API’s proximity verification was a challenging feature but turned out very nicely in the end.

# What we learned
We learned a lot about how to integrate a service such as Firebase with a Python library like Flask. In particular, the Flask framework had many issues that we needed to work around in order to properly authenticate using Firebase authentication and data. We also learned about QR code generation and the true interconnectedness of the web.

# What's next for Retro
In the future, we plan to expand Retro to be a product used across ALL hospitals, and to add a system of doctor credential validation for entering more rigorous and detailed patient information. In this way, Retro could truly revolutionize the current system of withholding medical records from patients and usher in a new era of transparency between patients and care providers.

<p align="middle">
    <img src="https://github.com/antz22/retro/blob/master/screenshots/qrcode.png" width="100%" style="margin:0; padding: 0">
</p>

# Built With

```
bootstrap
css3
firebase
flask
html5
javascript
mapbox
python
qrcode
realtime-database
```

changelog.txt
Version: 4.7.1
  - Fixed pricing table borders issue

Version: 4.7.0
  - Updated Bootstrap to version 5.1.3
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.6.1
  - Fixed the hero section, which was not displaying properly in mobile landscape mode

Version: 4.6.0
  - Updated Bootstrap to version 5.1.2
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.5.0
  - Fixed slider issue in testimonials and portfolio details sections

Version: 4.4.0
  - Updated Bootstrap to version 5.1.1
  - Updated all outdated third party vendor libraries to their latest versions
  - Improved and updated dev version gulp scripts

Version: 4.3.0
  - Updated Bootstrap to version 5.0.1
  - Updated all outdated third party vendor libraries to their latest versions
  - Fixed navigation links focus color

Version: 4.2.0
  - Updated Bootstrap to version 5.0.0 Final
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.1.0
  - Updated Bootstrap to version 5.0.0-beta3
  - Updated all outdated third party vendor libraries to their latest versions
  - Updated the PHP Email Form to V3.1

Version: 4.0.1
  - Updated Bootstrap to version 5.0.0-beta2
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.0.0
  - The template does not require jQuery anymore
  - Removed jQuery and all the jQuery plugins
  - The assets/js/main.js was rewritten completely with vanilla Javascript. No more jQuery code
  - Restructured the dev version for better development experience
  - Updated the PHP Email Form to V3.0 - No jQuery dependency. Added attachment support

Version: 3.0.0
  - Initial release with Bootstrap v5.0 Beta 1

Version: 2.2.0
  - Updated Bootstrap to version 4.5.3
  - Updated all outdated third party vendor libraries to their latest versions
  - Updated the PHP Email Form to v2.3
  - Other small fixes and improvements

Version: 2.1.0
  - Updated Bootstrap to version 4.5.0
  - Updated the PHP Email Form library to version 2.0 with reCaptcha support
  - Added inner-page.html page template
  - Added smooth scroll on page load with hash links in the url
  - Added navigation active state on scroll
  - Replaced the WOW animation on scroll library with AOS
  - Updated all outdated third party vendor libraries to their latest versions
  - Other small fixes and improvements
  
Version: 2.0.0
  - The template was rebuilt from scratch with the latest Bootstrap version (4.4.1)
  - Added SMPTP support for the contact form script (Pro)
  - Added NodeJS NPM Development version (Pro unlimited & Membership members)
  
Version: 1.0.0
- Initial Release


backend/static/js/main.js
/**
* Template Name: eStartup - v4.7.1
* Template URL: https://bootstrapmade.com/estartup-bootstrap-landing-page-template/
* Author: BootstrapMade.com
* License: https://bootstrapmade.com/license/
*/
(function() {
  "use strict";

  /**
   * Easy selector helper function
   */
  const select = (el, all = false) => {
    el = el.trim()
    if (all) {
      return [...document.querySelectorAll(el)]
    } else {
      return document.querySelector(el)
    }
  }

  /**
   * Easy event listener function
   */
  const on = (type, el, listener, all = false) => {
    let selectEl = select(el, all)
    if (selectEl) {
      if (all) {
        selectEl.forEach(e => e.addEventListener(type, listener))
      } else {
        selectEl.addEventListener(type, listener)
      }
    }
  }

  /**
   * Easy on scroll event listener 
   */
  const onscroll = (el, listener) => {
    el.addEventListener('scroll', listener)
  }

  /**
   * Navbar links active state on scroll
   */
  let navbarlinks = select('#navbar .scrollto', true)
  const navbarlinksActive = () => {
    let position = window.scrollY + 200
    navbarlinks.forEach(navbarlink => {
      if (!navbarlink.hash) return
      let section = select(navbarlink.hash)
      if (!section) return
      if (position >= section.offsetTop && position <= (section.offsetTop + section.offsetHeight)) {
        navbarlink.classList.add('active')
      } else {
        navbarlink.classList.remove('active')
      }
    })
  }
  window.addEventListener('load', navbarlinksActive)
  onscroll(document, navbarlinksActive)

  /**
   * Scrolls to an element with header offset
   */
  const scrollto = (el) => {
    let header = select('#header')
    let offset = header.offsetHeight

    if (!header.classList.contains('header-scrolled')) {
      offset -= 16
    }

    let elementPos = select(el).offsetTop
    window.scrollTo({
      top: elementPos - offset,
      behavior: 'smooth'
    })
  }

  /**
   * Toggle .header-scrolled class to #header when page is scrolled
   */
  let selectHeader = select('#header')
  if (selectHeader) {
    const headerScrolled = () => {
      if (window.scrollY > 100) {
        selectHeader.classList.add('header-scrolled')
      } else {
        selectHeader.classList.remove('header-scrolled')
      }
    }
    window.addEventListener('load', headerScrolled)
    onscroll(document, headerScrolled)
  }

  /**
   * Back to top button
   */
  let backtotop = select('.back-to-top')
  if (backtotop) {
    const toggleBacktotop = () => {
      if (window.scrollY > 100) {
        backtotop.classList.add('active')
      } else {
        backtotop.classList.remove('active')
      }
    }
    window.addEventListener('load', toggleBacktotop)
    onscroll(document, toggleBacktotop)
  }

  /**
   * Mobile nav toggle
   */
  on('click', '.mobile-nav-toggle', function(e) {
    select('#navbar').classList.toggle('navbar-mobile')
    this.classList.toggle('bi-list')
    this.classList.toggle('bi-x')
  })

  /**
   * Mobile nav dropdowns activate
   */
  on('click', '.navbar .dropdown > a', function(e) {
    if (select('#navbar').classList.contains('navbar-mobile')) {
      e.preventDefault()
      this.nextElementSibling.classList.toggle('dropdown-active')
    }
  }, true)

  /**
   * Scrool with ofset on links with a class name .scrollto
   */
  on('click', '.scrollto', function(e) {
    if (select(this.hash)) {
      e.preventDefault()

      let navbar = select('#navbar')
      if (navbar.classList.contains('navbar-mobile')) {
        navbar.classList.remove('navbar-mobile')
        let navbarToggle = select('.mobile-nav-toggle')
        navbarToggle.classList.toggle('bi-list')
        navbarToggle.classList.toggle('bi-x')
      }
      scrollto(this.hash)
    }
  }, true)

  /**
   * Scroll with ofset on page load with hash links in the url
   */
  window.addEventListener('load', () => {
    if (window.location.hash) {
      if (select(window.location.hash)) {
        scrollto(window.location.hash)
      }
    }
  });

  /**
   * Clients Slider
   */
  new Swiper('.screens-slider', {
    speed: 400,
    loop: true,
    autoplay: {
      delay: 5000,
      disableOnInteraction: false
    },
    slidesPerView: 'auto',
    pagination: {
      el: '.swiper-pagination',
      type: 'bullets',
      clickable: true
    },
    breakpoints: {
      320: {
        slidesPerView: 1,
        spaceBetween: 20
      },
      575: {
        slidesPerView: 2,
        spaceBetween: 20
      },
      640: {
        slidesPerView: 3,
        spaceBetween: 20
      },
      992: {
        slidesPerView: 4,
        spaceBetween: 20
      }
    }
  });

  /**
   * Initiate  glightbox 
   */
  const glightbox = GLightbox({
    selector: '.glightbox'
  });

  /**
   * Animation on scroll
   */
  window.addEventListener('load', () => {
    AOS.init({
      duration: 1000,
      easing: 'ease-in-out',
      once: true,
      mirror: false
    })
  });

})()

backend/main.py
from flask import Flask, redirect, url_for, render_template, request, flash
from datetime import date
import pyrebase
import qrcode
import json
import time
import io
import base64

BASE_URL = 'http://127.0.0.1:5000'

# Connect to Firebase Realtime Database
config = {
  "apiKey": "AIzaSyBGRn-c9Cl6kbROuWu44w5besGDNxsDM5I",
  "authDomain": "retro-c0350.firebaseapp.com",
  "projectId": "retro-c0350",
  "databaseURL": "https://retro-c0350-default-rtdb.firebaseio.com/",
  "storageBucket": "retro-c0350.appspot.com",
  "messagingSenderId": "112454151211",
  "appId": "1:112454151211:web:abd98d22947b2b37ba0e6e"
}

firebase = pyrebase.initialize_app(config)
auth = firebase.auth()
db = firebase.database()
storage = firebase.storage()

USER = {"idToken": None, "localId": None, "mapVerified": False}

app = Flask(__name__)

# Views
@app.route('/')
def index():
    return render_template('index.html')


@app.route('/signup', methods = ["POST", "GET"])
def signup():
    if request.method == "POST":
        result = request.form
        email = result["email"]
        password = result["password"]
        name = result["name"]
        date_of_birth = result["date_of_birth"]
        sex = result["sex"]

        # date of birth
        today = date.today()

        yr = int(date_of_birth[:4])

        age = today.year - yr
        
        if email is None or password is None or name is None or date_of_birth is None or sex is None:
            return {'message': 'Error missing information'}, 400
        try:
            auth.create_user_with_email_and_password(email, password)
            user = auth.sign_in_with_email_and_password(email, password)
            id = user['localId']
            url = f'{BASE_URL}/portal/{id}'

            # Firebase Realtime Database
            data = {
                "id": id,
                "name": name,
                "email": email,
                "date_of_birth": date_of_birth,
                "url": url,
                "sex": sex,
                "height": "N/A",
                "weight": "N/A",
                "blood_type": "N/A",
                "emergency_phone": "N/A",
                "emergency_email": "N/A",
                "past_conditions": {
                    "initial": "N/A",
                },
                "treatments": {
                    "initial": {
                        "hospital": "N/A",
                        "treatment": "N/A"
                    }
                },
                "blood_tests": "N/A",
                "age": age,
                "files": {
                    "initial": True,
                },
            }
            db.child("users").child(id).set(data)

            USER["idToken"] = user['idToken']
            USER["localId"] = user['localId']

            return redirect(url_for('personal_portal'))
        except:
            print('error creating user')
            return {'message': 'Error creating user'}, 400
    else:
        return render_template('signup.html')


@app.route('/login', methods = ["POST", "GET"])
def login():
    if request.method == "POST":
        result = request.form
        email = result["email"]
        password = result["password"]
        if email is None or password is None:
            return {'message': 'Error missing information'}, 400
        try:
            user = auth.sign_in_with_email_and_password(email, password)
            USER["idToken"] = user['idToken']
            USER["localId"] = user['localId']

            return redirect(url_for('personal_portal'))
        except:
            return {'message': 'Error logging in'}, 400
    else:
        return render_template('login.html')


@app.route('/map')
def map():
    return render_template('map.html')


@app.route('/personal-portal')
def personal_portal():
    db = firebase.database()
    if USER["idToken"] != None and USER["localId"] != None and USER["mapVerified"] == False:
        USER["mapVerified"] = True
        return redirect(url_for('map'))
    elif USER["idToken"] != None and USER["localId"] != None and USER["mapVerified"] == True:
        id = USER["localId"]
        user = db.child("users").child(id).get()
        user_data = user.val()
        name = user_data["name"]
        date_of_birth = user_data["date_of_birth"]
        email = user_data["email"]
        url = user_data["url"]
        sex = user_data["sex"]
        height = user_data["height"]
        weight = user_data["weight"]
        blood_type = user_data["blood_type"]
        age = user_data["age"]
        past_conditions = user_data["past_conditions"]
        treatments = user_data["treatments"]
        blood_tests = user_data["blood_tests"]
        emergency_phone = user_data["emergency_phone"]
        emergency_email = user_data["emergency_email"]

 

        data = io.BytesIO()
        img = getQRCode(id)
        img.save(data, "JPEG")
        encoded_img_data = base64.b64encode(data.getvalue())

        return render_template('dashboard/pages/dashboard.html', img_data=encoded_img_data.decode('utf-8'), name=name, email=email, url=url, id=id, age=age, sex=sex, height=height, weight=weight, blood_type=blood_type, past_conditions=past_conditions, treatments=treatments, blood_tests=blood_tests, emergency_phone=emergency_phone, emergency_email=emergency_email)
    else:
        return redirect(url_for('login'))


@app.route('/portal/<id>')
def portal(id):
    db = firebase.database()
    user = db.child("users").child(id).get()
    user_data = user.val()

    id = user_data["id"]
    name = user_data["name"]
    date_of_birth = user_data["date_of_birth"]
    email = user_data["email"]
    url = user_data["url"]
    sex = user_data["sex"]
    height = user_data["height"]
    weight = user_data["weight"]
    blood_type = user_data["blood_type"]
    age = user_data["age"]
    past_conditions = user_data["past_conditions"]
    treatments = user_data["treatments"]
    blood_tests = user_data["blood_tests"]
    emergency_phone = user_data["emergency_phone"]
    emergency_email = user_data["emergency_email"]
    

    data = io.BytesIO()
    img = getQRCode(id)
    img.save(data, "JPEG")
    encoded_img_data = base64.b64encode(data.getvalue())

    return render_template('dashboard/pages/dashboard.html', img_data=encoded_img_data.decode('utf-8'), name=name, email=email, url=url, id=id, age=age, sex=sex, height=height, weight=weight, blood_type=blood_type, past_conditions=past_conditions, treatments=treatments, blood_tests=blood_tests, emergency_phone=emergency_phone, emergency_email=emergency_email)


@app.route('/add-entry', methods = ["POST", "GET"])
def add_entry():


    if request.method == "POST":
        result = request.form
        height = result["height"]
        weight = result["weight"]
        blood_type = result["blood_type"]
        blood_tests = result["blood_tests"]
        emergency_phone = result["emergency_phone"]
        emergency_email = result["emergency_email"]
        if height is None or weight is None or blood_type is None or blood_tests is None or emergency_phone is None or emergency_email is None:
            return {'message': 'Error missing information'}, 400
        try:
            id = USER['localId']

            # Firebase Realtime Database
            data = {
                "height": height,
                "weight": weight,
                "blood_type": blood_type,
                "blood_tests": blood_tests,
                "emergency_phone": emergency_phone,
                "emergency_email": emergency_email
            }
            db.child("users").child(id).update(data)


            return redirect(url_for('personal_portal'))
        except:
            print('error updating user')
            return {'message': 'Error updating user'}, 400
    else:
        return render_template('add_entry.html')



@app.route('/add-file', methods = ["POST", "GET"])
def add_file():
    db = firebase.database()
    id = USER["localId"]
    if id == None:
        return redirect(url_for('login'))

    if request.method == "POST":
        file = request.files['file']
        if file.filename == '':
            flash('No selected file')
            return {'message': 'Error missing information'}, 400
        # UPLOAD
        storage_rel_url = f"files/{id}/{time.time()}.pdf"
        storage.child(storage_rel_url).put(file)

        timestamp = str(time.time())[:9]
        
        try:

            # Firebase Realtime Database
            data = {
                timestamp: f"{storage_rel_url}",
            }

            if db.child("users").child(id).child("files").child("initial").get() == None:
                print('hi')
                db.child("users").child(id).child("files").child("initial").remove()
                db.child("users").child(id).child("files").set(data)
            else:
                db.child("users").child(id).child("files").update(data)

            return redirect(url_for('personal_portal'))
        except Exception as e:
            print(e)
            print('error updating user')
            return {'message': 'Error updating user'}, 400
    else:
        return render_template('add_file.html')


@app.route('/view-file')
def view_file():
    db = firebase.database()
    id = USER["localId"]
    if id == None:
        return redirect(url_for('login'))

    user = db.child("users").child(id).get()
    user_data = user.val()

    file_urls = user_data["files"]
    display_urls = []

    for key, url in file_urls.items():
        full_url = storage.child(url).get_url(USER["idToken"])
        display_urls.append(full_url)
    
    data = io.BytesIO()
    img = getQRCode(id)
    img.save(data, "JPEG")
    encoded_img_data = base64.b64encode(data.getvalue())

    return render_template('view-pdfs.html', urls=display_urls, img_data=encoded_img_data.decode('utf-8'))
    

def getQRCode(id):
    img = qrcode.make('{}/portal/{}'.format(BASE_URL, id))
    return img

if __name__ == '__main__':
    app.run(debug=True)